### PR TITLE
MPP-3932: Disable masks on complaints, round 2

### PR DESCRIPTION
**Pull request by jwhitlock**
This replaces #5114, and goes a lot farther.

This rewrites `_handle_complaint`, moving most of the logic into three new helper functions:

* `_get_complaint_data` gets the relevant data out of the complaint notification, with lots of checks and logging if the message is not what we're expecting.
* `_gather_complainers` cross-checks the sender and receiver data against the database, getting the related users and masks (`RelayAddress`, `DomainAddress`). It also checks for the developer "simulate_complaint" mode, and swaps the real user email back in.
* `_reduce_future_complaints` takes the actions to reduce future complaints.
  - On the first complaint, turns on the `auto_block_spam` flag to drop emails that AWS SES identifies as spam (same as before)
  - On the second complaint, it turns the associated mask to block emails and sends an email to the user. This has been altered to look at the `From:` header data, which has the mask, rather than the `source` field, which is our reply address.
  - On future complaints (`how?`), it does nothing but takes note that it did nothing (new)

This leaves `_handle_complaint` to log the complaint in developer mode, log what happened for each detected Relay user, and emit a counter.

I also extended the developer mode to correctly work for domain masks, added `_get_address_if_exists`, removed a `mypy` override, and fixed Heroku Redis... This PR can definitely be broken into smaller bits. 

# How to test:

The test suite has been expanded to cover the new code. There are lots of new tests.

I pushed this branch to Heroku for manual testing. There are two issues ~I've seen~ that are now fixed on Heroku:

* Each forwarded email appears to result in two complaints - Misconfiguration of SES on dev server
* In the email "Firefox Relay⁩ has deactivated one of your email masks", there's some extra `???` and the links do not work. - Fixed the HTML template

Here's the test process:

* Deploy this branch
* Send a test email to a mask. It should be forwarded.
* Load the Papertrail app in Heroku. Filter on `program:"app/worker.1"`. You'll see logs related to your email.
* Add the `developer_mode` flag to your Heroku user. You can run a Heroku bash shell:
  `heroku run --app fx-private-relay bash`
  then run
  `./manage.py waffle_flag developer_mode --append --user "<your_real_email@example.com>"`
* Create a random mask, give it the description "DEV:simulate_complaint" (no spaces)
* Send an email to that mask
  - The email is _not_ delivered to you, but instead to the [AWS SES Complaint Simulator address](https://docs.aws.amazon.com/ses/latest/dg/send-an-email-from-console.html#send-email-simulator)
  - You should see several logs with `"msg": "_handle_received: developer_mode"`. The `"notification_gza85"` fields can be combined and passed to `emails.utils.decode_dict_gza85` to get the original delivery notification.
  - After a few more messages, you should see a log with `"msg": "_handle_complaint: developer mode"`. The `"notification_gza85"` fields can be combined and passed to `emails.utils.decode_dict_gza85` to get the original complaint notification.
  - The next log is `"msg": "complaint_notification"`, and has the standard log for a complaint. The `"relay_action"` should be `"auto_block_spam"`
  - There will be other logs as well, but these are the important ones.
* Send a second email to that mask
  - The logs repeat, with `"relay_action"` as `no_action` since `auto_block_spam` is already on.
* Add the `disable_mask_on_complaint` flag to your Heroku user. You can run a Heroku bash shell:
  `heroku run --app fx-private-relay bash`
  then run
  `./manage.py waffle_flag disable_mask_on_complaint --append --user "<your_real_email@example.com>"`
* Send a third email to that mask.
  - The logs repeat. The `"relay_action"` is `"disable_mask"` (and then `no_action` if there is a second complaint)
* Check your email after a minute
  - You should get the email "Firefox Relay⁩ has deactivated one of your email masks." to your real address. The links should go to the email dashboard.
* Go to the Email Dashboard, refreshing if needed (such as going to FAQ and back to the dashboard)
  - The mask is now set to Blocking

### CODE REVIEW

File: emails/views.py [hunk 1]
```diff
@@ -1709,73 +1711,26 @@ def _send_disabled_mask_for_spam_email(
             message=msg,
         )
     except ClientError as e:
-        logger.error("reply_not_allowed_ses_client_error", extra=e.response["Error"])
-    incr_if_enabled("free_user_reply_attempt", 1)
```
Comment (jwhitlock) [inline emails/views.py:1713]: These were copied from the `reply_not_allowed` email function, I changed them.

File: emails/views.py [hunk 2]
```diff
@@ -1709,73 +1711,26 @@ def _send_disabled_mask_for_spam_email(
...
-            ObjectDoesNotExist,
-            RelayAddress.DoesNotExist,
-            DomainAddress.DoesNotExist,
-        ):
-            logger.error(
-                "Received a complaint from a destination address that does not match "
-                "a Relay address.",
-            )
+        logger.error("send_disabled_mask_ses_client_error", extra=e.response["Error"])
+    incr_if_enabled("send_disabled_mask_email", 1)
```
Comment (jwhitlock) [inline emails/views.py:1715]: this is the new log error and metric for sending disabled mask emails.

File: emails/views.py [hunk 3]
```diff
@@ -1709,73 +1711,26 @@ def _send_disabled_mask_for_spam_email(
...
 
 def _handle_complaint(message_json: AWS_SNSMessageJSON) -> HttpResponse:
     """
     Handle an AWS SES complaint notification.
 
-    Sets the user's auto_block_spam flag to True.
+    This looks for Relay users in the complainedRecipients (real email address)
+    and the From: header (mask address). We expect both to match the same Relay user,
+    and return a 200. If one or the other do not match, a 404 is returned, and errors
+    may be logged.
```
Comment (jwhitlock) [inline emails/views.py:1725]: GitHub diff gets confused here. Basically everything below here has changed.

File: emails/views.py [hunk 4]
```diff
@@ -1709,73 +1711,26 @@ def _send_disabled_mask_for_spam_email(
             message=msg,
         )
     except ClientError as e:
-        logger.error("reply_not_allowed_ses_client_error", extra=e.response["Error"])
-    incr_if_enabled("free_user_reply_attempt", 1)
-
-
-def _disable_masks_for_complaint(message_json: dict, user: User) -> None:
-    """
-    See https://docs.aws.amazon.com/ses/latest/dg/notification-contents.html#mail-object
```
Comment (jwhitlock) [inline emails/views.py:1718]: This function has been absorbed into `_reduce_future_complaints`

File: pyproject.toml [hunk 5]
```diff
@@ -58,7 +58,6 @@ module = [
     "django_ftl",
     "django_ftl.bundles",
     "google.*",
-    "google_measurement_protocol",
```
Comment (jwhitlock) [inline pyproject.toml:61]: This is the only line I meant to change in here. The rest were done by my editor's TOML formatter.

**Review by jwhitlock — COMMENTED** [review]
Whoa this is big for just 3 files. You should probably skim `emails/tests/views_tests.py`, and focus on `emails/views.py`. Here's some comments that might help.

**Comment by jwhitlock** [comment]
There are indeed 2 complaints for each time the simulator is used on dev. For one pair, the first has a `timestamp` and `arrivalDate` of `2024-10-16T23:15:27.100Z`. For the second, the `timestamp` was `"timestamp": "2024-10-16T23:15:27.000Z` (0.1 seconds earlier) and `arrivalDate` was `2024-10-16T23:15:27.167Z` (0.06 seconds later). The other differences

* The first had `eventType: Complaint` and the second had `notificationType: Complaint`
* The second had `mail.sourceIP` and `mail.callerIdentity`, the first did not
* The first had `mail.commonHeaders.messageId`, the second did not
* The first had `mail.tags` with several items list `ses:source-tls-version`, the second did not.

I'm going to see if there is a similar pattern in production complaints. If not, this may be a dev misconfiguration, or a quirk of the simulator. It looks like we might want to ignore either `eventType` or `notificationType`.

**Comment by jwhitlock** [comment]
No, production does not have this pattern. 

Digging in, there are two ways to monitor complaints:

* Set up event publishing with a [configuration set](https://docs.aws.amazon.com/ses/latest/dg/event-publishing-create-configuration-set.html), and specify this configuration set when sending email. This will send `"eventType": "Complaint"` objects to your configured destination (SNS -> SQS for us)
* Set up a [domain identity](https://docs.aws.amazon.com/ses/latest/dg/configure-sns-notifications.html#configure-feedback-notifications-console) for your email sending domain, and configure Email feedback forwarding. This will send `"notificationType": "Complaint"` objects to your configured destination (SNS -> SQS for us).

We have both configured on the development server. Looking at my notes, it is possible I enabled this, as I was looking into complaint formats in May 2022. I turned off email feedback forwarding of complaints for our dev domain identity. This will also turn off duplicate bounce notifications and the unhandled delivery notifications.

On stage, we have not configured the SNS topics for the email feedback forwarding, no changes there.

File: emails/views.py [hunk 6]
```diff
@@ -1784,84 +1739,293 @@ def _handle_complaint(message_json: AWS_SNSMessageJSON) -> HttpResponse:
...
+    complaint_data: RawComplaintData,
+) -> tuple[list[Complainer], int]:
+    """Fetch Relay Users and masks from the complaint data"""
+
+    users: dict[int, Complainer] = {}
+    unknown_complainer_count = 0
+    for email_address, extra_data in complaint_data.complained_recipients:
+        local, domain = email_address.split("@", 1)
+
+        # For developer mode complaint simulation, swap with developer's email
```
Comment (groovecoder) [inline emails/views.py:1891]: nit (non-blocking): this code seems specific to AWS complaint simulator, but not necessarily to the "developer mode" controlled by the mask label?

If it is, can we link from here to a doc/comment about the "developer mode"?

```suggestion
        # For AWS complaint simulation, swap with developer's email
```

**Review by groovecoder — COMMENTED** [review]
Great update! I haven't looked at the test file yet, and it looks like you updated it again since I started the review, but here's my first few comments.

File: emails/tests/views_tests.py [hunk 7]
```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
...
-        assert log_complaint == expected_log_complaint
+        assert log_complaint == simulator_complaint_message
 
-        record = logs.records[1]
-        assert record.msg == "complaint_notification"
+        assert rec2.msg == "complaint_notification"
+
+    @override_flag("developer_mode", active=True)
+    def test_complaint_developer_mode_missing_mask(self):
+        """If the simulator email can not be turned into a mask, it is not changed."""
```
Comment (groovecoder) [inline emails/tests/views_tests.py:1357]: sugguestion (non-blocking): add a little more detail here? I don't understand what "simulator email can not be turned into a mask" means.

File: emails/tests/views_tests.py [hunk 8]
```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
...
+        assert log_complaint == simulator_complaint_message
+
+        assert rec2.msg == "complaint_notification"
+
+        (err_log,) = error_logs.records
+        assert err_log.msg == "_gather_complainers: unknown complainedRecipient"
+
+    @override_flag("developer_mode", active=True)
+    def test_complaint_developer_mode_domain_address(self):
+        """If the simulator email can not be turned into a mask, it is not changed."""
```
Comment (groovecoder) [inline emails/tests/views_tests.py:1394]: same sugguestion (non-blocking): add a little more detail here? I don't understand what "simulator email can not be turned into a mask" means.

File: emails/tests/views_tests.py [hunk 9]
```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
...
+
+        (rec1, rec2) = logs.records
+        assert rec1.msg == "_handle_complaint: developer_mode"
+        assert getattr(rec1, "mask_id") == domain_address.metrics_id
+        assert getattr(rec1, "dev_action") == "log"
+
+        assert rec2.msg == "complaint_notification"
+
+    def test_complaint_no_complained_recipients_error_logged(self):
+        """Without complained recipients, an error is logged, but matches on From"""
```
Comment (groovecoder) [inline emails/tests/views_tests.py:1429]: suggestion (non-blocking): I had to read thru the code and test fixture to understand what this test was doing.

```suggestion
        """
        Without complainedRecipients, we can still match the user on From:,
        perform the appropriate action, and log an error
        """
```

File: emails/tests/views_tests.py [hunk 10]
```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
...
+
+        (err_log,) = error_logs.records
+        assert err_log.msg == "_get_complaint_data: Unexpected message"
+        assert getattr(err_log, "missing_key") == "complainedRecipients"
+        assert getattr(err_log, "found_keys") == (
+            "arrivalDate,complaintFeedbackType,feedbackId,timestamp,userAgent"
+        )
+
+    def test_complaint_empty_complained_recipients_error_logged(self):
+        """With empty complained recipients, an error is logged, but matches on From"""
```
Comment (groovecoder) [inline emails/tests/views_tests.py:1459]: same suggestion (non-blocking):

```suggestion
        """
        With complainedRecipients empty, we can still match the user on From:,
        perform the appropriate action, and log an error
        """
```

File: emails/tests/views_tests.py [hunk 11]
```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
...
+
+        self.user.profile.refresh_from_db()
+        assert self.user.profile.auto_block_spam is True
+        self.mock_ses_client.send_raw_email.assert_not_called()
+
+        (err_log,) = error_logs.records
+        assert err_log.msg == "_get_complaint_data: Empty complainedRecipients"
+
+    def test_complaint_wrong_complained_recipients_error_logged(self):
+        """With an unexpected object in complained recipients, an error is logged"""
```
Comment (groovecoder) [inline emails/tests/views_tests.py:1476]: same suggestion (non-blocking):

```suggestion
        """
        With unexpected value in complainedRecipients,
        we can still match the user on From:, perform the appropriate action,
        and log an error
        """
```

File: emails/tests/views_tests.py [hunk 12]
```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
...
+        assert self.user.profile.auto_block_spam is True
+        self.mock_ses_client.send_raw_email.assert_not_called()
+
+        (err_log,) = error_logs.records
+        assert err_log.msg == "_get_complaint_data: Unexpected message"
+        assert getattr(err_log, "missing_key") == "emailAddress"
+        assert getattr(err_log, "found_keys") == "foo"
+
+    def test_complaint_no_mail_error_logged(self):
+        """If the mail object is missing, an error is logged"""
```
Comment (groovecoder) [inline emails/tests/views_tests.py:1495]: similar suggestion (non-blocking):

```suggestion
        """
        With no "mail" object in the complaint, we can still match the user on
        complainedRecipients, perform the appropriate action, and log an error.
        """
```

File: emails/tests/views_tests.py [hunk 13]
```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
...
+        assert self.user.profile.auto_block_spam is True
+        self.mock_ses_client.send_raw_email.assert_not_called()
+
+        (err_log,) = error_logs.records
+        assert err_log.msg == "_get_complaint_data: Unexpected message"
+        assert getattr(err_log, "missing_key") == "mail"
+        assert getattr(err_log, "found_keys") == "complaint,notificationType"
+
+    def test_complaint_no_common_headers_error_logged(self):
+        """If the mail object is missing, an error is logged"""
```
Comment (groovecoder) [inline emails/tests/views_tests.py:1514]: similar suggestion (non-blocking) and correction (blocking):

```suggestion
        """
        With no "commonHeaders" in the mail object of the complaint, we can still match
        the user on complainedRecipients, perform the appropriate action,
        and log an error.
        """
```

File: emails/tests/views_tests.py [hunk 14]
```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
...
+
+        (err_log,) = error_logs.records
+        assert err_log.msg == "_get_complaint_data: Unexpected message"
+        assert getattr(err_log, "missing_key") == "commonHeaders"
+        assert getattr(err_log, "found_keys") == (
+            "destination,headers,headersTruncated,messageId,source,timestamp"
+        )
+
+    def test_complaint_no_from_header_error_logged(self):
+        """If the mail object is missing, an error is logged"""
```
Comment (groovecoder) [inline emails/tests/views_tests.py:1544]: similar suggestion (non-blocking) and correction (blocking):

```suggestion
        """
        With no "from" in the commonHeaders of the complaint, we can still match
        the user on complainedRecipients, perform the appropriate action,
        and log an error.
        """
```

File: emails/tests/views_tests.py [hunk 15]
```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
...
+        assert self.user.profile.auto_block_spam is True
+        self.mock_ses_client.send_raw_email.assert_not_called()
+
+        (err_log,) = error_logs.records
+        assert err_log.msg == "_get_complaint_data: Unexpected message"
+        assert getattr(err_log, "missing_key") == "from"
+        assert getattr(err_log, "found_keys") == "subject,to"
+
+    def test_complaint_unknown_complained_recipient_404(self):
+        """If the recipient is unknown but sender is mask, return 404"""
```
Comment (groovecoder) [inline emails/tests/views_tests.py:1563]: similar suggestion (non-blocking):

```suggestion
        """
        With unknown emailAddress in complainedRecipients, we can still match the 
        user on From:, perform the appropriate action, and log an error.
        """
```

File: emails/tests/views_tests.py [hunk 16]
```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
...
+
+        (err_log,) = error_logs.records
+        assert err_log.msg == "_gather_complainers: unknown complainedRecipient"
+
+    def test_complaint_unknown_complained_recipient_two_masks_404(self):
+        """
+        If the recipient is unknown but two senders are masks, return 404
+
+        Also log the weirdness of two senders that match the user.
+        """
```
Comment (groovecoder) [inline emails/tests/views_tests.py:1595]: similar suggestion (non-blocking):

```suggestion
        """
        With unknown emailAddress in complainedRecipients, and 2 masks in From:,
        we can still match the user on From:, perform the appropriate action,
        and log an error.
        """
```

File: emails/tests/views_tests.py [hunk 17]
```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
...
+        assert getattr(info_log, "user_match") == "found"
+        assert getattr(info_log, "found_in") == "from_header"
+        assert getattr(info_log, "relay_action") == "auto_block_spam"
+
+        (err_log1, err_log2) = error_logs.records
+        assert err_log1.msg == "_gather_complainers: unknown complainedRecipient"
+        assert err_log2.msg == "_gather_complainers: no complainer, multi-mask"
+
+    def test_complaint_unknown_from_header_404(self):
+        """If the recipient is known but sender is unknown, return 404"""
```
Comment (groovecoder) [inline emails/tests/views_tests.py:1627]: similar suggestion (non-blocking):

```suggestion
        """
        With unknown address in From:, we can still match the user on complainedRecipients,
        perform the appropriate action, and log an error.
        """
```

File: emails/tests/views_tests.py [hunk 18]
```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
...
+        assert info_log.msg == "complaint_notification"
+        assert getattr(info_log, "user_match") == "found"
+        assert getattr(info_log, "found_in") == "complained_recipients"
+        assert getattr(info_log, "relay_action") == "auto_block_spam"
+
+        (err_log,) = error_logs.records
+        assert err_log.msg == "_gather_complainers: unknown mask, maybe deleted?"
+
+    def test_complaint_complained_recipient_twice_log_error(self):
+        """If a complainer appears twice, log the weirdness."""
```
Comment (groovecoder) [inline emails/tests/views_tests.py:1655]: similar suggestion (non-blocking):

```suggestion
        """
        With a complainer in complainedRecipients twice, we can still match the user,
        perform the appropriate action, and log an error.
        """
```

File: emails/tests/views_tests.py [hunk 19]
```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
...
+
+        self.user.profile.refresh_from_db()
+        assert self.user.profile.auto_block_spam is True
+        self.mock_ses_client.send_raw_email.assert_not_called()
+
+        (err_log,) = error_logs.records
+        assert err_log.msg == "_gather_complainers: complainer appears twice, discarded"
+
+    def test_complaint_from_header_twice_log(self):
+        """If an email appears in a from header twice, log the weirdness."""
```
Comment (groovecoder) [inline emails/tests/views_tests.py:1674]: similar suggestion (non-blocking):

```suggestion
        """
        With a mask in From: twice, we can still match the user,
        perform the appropriate action, and log an error.
        """
```

File: emails/tests/views_tests.py [hunk 20]
```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
...
+
+        self.user.profile.refresh_from_db()
+        assert self.user.profile.auto_block_spam is True
+        self.mock_ses_client.send_raw_email.assert_not_called()
+
+        (err_log,) = error_logs.records
+        assert err_log.msg == "_gather_complainers: mask appears twice"
+
+    def test_complaint_from_stranger_is_404(self):
+        """If no Relay users match, log the complaint."""
```
Comment (groovecoder) [inline emails/tests/views_tests.py:1693]: similar suggestion (non-blocking):

```suggestion
        """
        If we can't find a Relay user via complainedRecipients or From:,
        all we can do is log an error.
        """
```

File: emails/tests/views_tests.py [hunk 21]
```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
...
+            {"emailAddress": "complaint+missing@simulator.amazonses.com"}
+        ]
+        complaint_body = {"Message": json.dumps(simulator_complaint_message)}
+
+        with (
+            self.assertLogs(INFO_LOG) as logs,
+            self.assertLogs(ERROR_LOG) as error_logs,
+        ):
+            response = _sns_notification(complaint_body)
+        assert response.status_code == 404
```
Comment (groovecoder) [inline emails/tests/views_tests.py:1370]: question (blocking): It looks like, unlike the code in `process_emails_from_sqs`, this complaint-handling code is run in the context of an HTTP callback request from AWS SNS. What does SNS do if it receives a 404 from Relay?

File: emails/tests/views_tests.py [hunk 22]
```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
...
+            {"emailAddress": "unknown@somewhere.example.com"}
+        ]
+        complaint_body = {"Message": json.dumps(complaint_msg)}
+
+        with (
+            self.assertLogs(INFO_LOG) as info_logs,
+            self.assertLogs(ERROR_LOG) as error_logs,
+        ):
+            response = _sns_notification(complaint_body)
+        assert response.status_code == 404
```
Comment (groovecoder) [inline emails/tests/views_tests.py:1575]: same question (blocking): re: 404 effect on AWS

File: emails/tests/views_tests.py [hunk 23]
```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
...
+        ]
+        complaint_msg["mail"]["commonHeaders"]["from"].append(second_mask.full_address)
+        complaint_body = {"Message": json.dumps(complaint_msg)}
+
+        with (
+            self.assertLogs(INFO_LOG) as info_logs,
+            self.assertLogs(ERROR_LOG) as error_logs,
+        ):
+            response = _sns_notification(complaint_body)
+        assert response.status_code == 404
```
Comment (groovecoder) [inline emails/tests/views_tests.py:1610]: same question (blocking): re: 404 effect on AWS

File: emails/tests/views_tests.py [hunk 24]
```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
...
+            "unknown@somewhere.example.com"
+        ]
+        complaint_body = {"Message": json.dumps(complaint_msg)}
+
+        with (
+            self.assertLogs(INFO_LOG) as info_logs,
+            self.assertLogs(ERROR_LOG) as error_logs,
+        ):
+            response = _sns_notification(complaint_body)
+        assert response.status_code == 404
```
Comment (groovecoder) [inline emails/tests/views_tests.py:1639]: same question (blocking): re: 404 effect on AWS

File: emails/tests/views_tests.py [hunk 25]
```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
...
+        ]
+        complaint_msg["mail"]["commonHeaders"]["from"] = ["sender@stranger.example.com"]
+        complaint_body = {"Message": json.dumps(complaint_msg)}
+
+        with (
+            self.assertLogs(INFO_LOG) as info_logs,
+            self.assertLogs(ERROR_LOG) as error_logs,
+        ):
+            response = _sns_notification(complaint_body)
+        assert response.status_code == 404
```
Comment (groovecoder) [inline emails/tests/views_tests.py:1706]: same question (blocking): re: 404 effect on AWS

**Review by groovecoder — CHANGES_REQUESTED** [review]
All comments are "dupes" of 4 main issues:

suggestion (non-blocking): When I woke up this morning I had to remember what the new "developer mode" did. It would be great to copy some of [the explanation from that PR](https://github.com/mozilla/fx-private-relay/pull/5090#issue-2581799332) into a more permanent doc or comment in the code-base somewhere.

suggestion (non-blocking): Similarly, a lot of the doc-blocks for the new tests assume high-context knowledge, which we may not have next time we have to touch this code. So I made some suggestions for updating the doc-blocks to add more context.

correction (blocking): A couple of the test doc-blocks have copy-pasted incorrect text.

~question (blocking): It looks like, unlike the code in `process_emails_from_sqs`, this complaint-handling code is run in the context of an HTTP callback request from AWS SNS. What does SNS do if it receives a 404 from Relay?~ [Answer](https://github.com/mozilla/fx-private-relay/pull/5115#discussion_r1806855301)

File: emails/tests/views_tests.py (hunk 21 above)
Comment (jwhitlock) [reply emails/tests/views_tests.py:1370]: The only documentation I can find is point 5 on https://docs.aws.amazon.com/sns/latest/dg/SendMessageToHttp.prepare.html:

> 5. Make sure that your endpoint responds to the HTTP POST message from Amazon SNS with the appropriate status code. The connection will time out in approximately 15 seconds. If your endpoint does not respond before the connection times out, or if your endpoint returns a status code outside the range of 200–4xx, Amazon SNS will consider the delivery of the message as a failed attempt.

This suggests a 404 would be treated as a successful delivery. If logging is setup, the status code could be used to see how many complaints did not match current Relay users or masks.

File: emails/tests/views_tests.py (hunk 22 above)
Comment (jwhitlock) [reply emails/tests/views_tests.py:1575]: 404 is treated as a successful delivery

File: emails/tests/views_tests.py (hunk 24 above)
Comment (jwhitlock) [reply emails/tests/views_tests.py:1639]: 404 is treated as a successful delivery

File: emails/tests/views_tests.py (hunk 21 above)
Comment (groovecoder) [reply emails/tests/views_tests.py:1370]: Thanks for digging that up! In this case I think `404` is fine.

File: emails/tests/views_tests.py (hunk 7 above)
Comment (jwhitlock) [reply emails/tests/views_tests.py:1362]: I'm going to move these "cover all the edge cases" tests to a new class, so `ComplaintHandlingTest` is more like integration tests.

File: emails/views.py [hunk 26]
```diff
@@ -1784,84 +1739,293 @@ def _handle_complaint(message_json: AWS_SNSMessageJSON) -> HttpResponse:
...
+    else:
+        raw_from_addresses = []
+    from_addresses = [parseaddr(addr)[1] for addr in raw_from_addresses]
+
+    feedback_type, _ = get_or_log("complaintFeedbackType", complaint, str)
+
+    # Only present when destination is on account suppression list
+    subtype = complaint.get("complaintSubType", "")
+    # Only present for feedback reports
+    user_agent = complaint.get("userAgent", "")
```
[the remark this answers is not in the document]
Comment (jwhitlock) [reply emails/views.py:1865]: Nope. The agent is the mail provider's agent. All complaints from yahoo.com, ymail.com, yahoo.com.mx, yahoo.de, etc. have agent `Yahoo!-Mail-Feedback/2.0`. A lot use `ReturnPathFBL/2.0`. Microsoft leaves it blank. See the link "Relay Bounces and Complaints v2" on the Relay Engineering page on Confluence.

File: emails/tests/views_tests.py (hunk 8 above)
Comment (jwhitlock) [reply emails/tests/views_tests.py:1401]: I'm moving these tests of `_gather_complainers` out of `ComplaintHandlingTest` into a focused test case.

**Comment by jwhitlock** [comment]
@groovecoder, this should be ready for recheck. 

Main changes:

* Added `docs/developer_mode.md`
* Use the metrics ID in the complaint simulator email, like `complaint+R123@simulator.amazonses.com`. Added `_get_mask_by_metrics_id` function and `GetMaskByMetricsIdTest` test case.
* Added `GetComplaintDataTest` test case to test `_get_complaint_data`, and `GatherComplainersTest` test case to test `_gather_complainers`. These now have "happy path" tests, as well as the edge case tests to get full coverage. This makes these tests more focused and the test data easier to understand. I've moved the similar tests out of `ComplaintHandlingTest`, leaving it to test the most important end-to-end test cases. I could do something similar for `_reduce_further_complaints`, but that seemed unneeded.

Smaller changes:

* Added a `log_group_id` to the developer_mode log, to make it a bit easier to get all the logs for a long notification.
* Updated docstrings and comments as suggested
* Rebase on main

File: emails/views.py [hunk 27]
```diff
@@ -1784,84 +1738,331 @@ def _handle_complaint(message_json: AWS_SNSMessageJSON) -> HttpResponse:
...
+    users: dict[int, Complainer] = {}
+    unknown_complainer_count = 0
+    for email_address, extra_data in complaint_data.complained_recipients:
+        local, domain = email_address.split("@", 1)
+
+        # If the complainer is the AWS SES complaint simulation, assume that
+        # it was send by a user with the developer_mode flag. Look for
+        # a mask that matches the embedded mask metrics_id, and use
+        # the related user's email instead of the AWS simulator address.
+        # See docs/developer_mode.md
```
Comment (groovecoder) [inline emails/views.py:1912]: suggestion (blocking): Can we rename the doc file to `docs/developer_mode.md` like this comment says? That will help future ENGRs (including us) find the doc file when searching the repo for the name of this flag.

**Review by groovecoder — CHANGES_REQUESTED** [review]
Looks good! 1 small change to the doc file name to make it easier to find for future ENGRs (including us).

File: emails/views.py (hunk 27 above)
Comment (jwhitlock) [reply emails/views.py:1912]: Yes, nice catch!

**Review by groovecoder — APPROVED** [review]
